### ISSUE TITLE
CCR auto-follow can leave unfollowed indices during master failovers 

### ISSUE THREAD

**Issue by fcofdez** [elastic/elasticsearch#86654]
Today when the cluster is unstable and there are master failovers while
some new leader indices match an auto-follow pattern it is possible that
the following index end up in a state where it does not pull changes from
the leader index or it is considered as an already following index when it 
is not followed.

One scenario where this is possible is after the following index is recovered
from the leader index in:
https://github.com/elastic/elasticsearch/blob/c7dc89f3cd86dbc9ad11c2f831c63651053e6e4a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java#L269-L270

Eventhough that end up calling a `AcknowledgedTransportMasterNodeAction`, it
uses the default timeout (30s) meaning that if there's a failure for more than
(30s) the listener just logs the failure instead of retrying or informing back
to the auto-follow coordinator, see:

https://github.com/elastic/elasticsearch/blob/c7dc89f3cd86dbc9ad11c2f831c63651053e6e4a/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java#L244-L255

**Comment by henningandersen** [comment]
I wonder if we should consider this a problem with auto-following or a problem with "put-follow" instead? Ideally, if the put-follow creates the index, we should also eventually start following.

**Comment by fcofdez** [comment]
I agree but I suspect that we have some cases where we might end up trying to "auto-follow" an already followed index multiple times during master failovers, this needs some tests to prove it.

**Comment by DaveCTurner** [comment]
By my reading if we do a resume-follow action on a shard that's already following then it will fail with a `ResourceAlreadyExistsException` rather than creating a duplicate follower task. It is tricky tho, I don't know that there's a way to distinguish a shard which failed to create the initial follower task from one whose following was subsequently paused.  We might need to add a flag to its index metadata and then do a single cluster state update which creates the follower task and flips the flag.

**Comment by BrianRothermich** [comment]
It looks like there is another possible path to this situation that does **not** involve a master failover as observed in this [SDH](https://github.com/elastic/sdh-elasticsearch/issues/10106)

In a captured occurrence:

- CCR started restoring the follower index at `t=0`.
- The follower primary started and the cluster became green at approximately `t=16.2s`.
- `RestoreService` reported the restore as `SUCCESS` at `t=16.4s`.
- The completed restore entry was removed from cluster state at `t=16.5s`.
- The put-follow callback completed at `t=28.7s` with:

```text
PutFollowAction.Response{
  followIndexCreated=true,
  followIndexShardsAcked=false,
  indexFollowingStarted=false
}
```

All relevant cluster-state publications came from the same master and term, ruling out master failover.

The apparent race is:

1. `RestoreService` publishes the restore cluster-state update. Its listener records cluster-state completion [here](https://github.com/elastic/elasticsearch/blob/v9.4.1/server/src/main/java/org/elasticsearch/snapshots/RestoreService.java#L1789-L1797)

2. The same operation also starts an allocation reroute and supplies a separate completion listener [here](https://github.com/elastic/elasticsearch/blob/v9.4.1/server/src/main/java/org/elasticsearch/snapshots/RestoreService.java#L1591-L1599)

3. These two listeners are combined using `AllocationActionListener`, which invokes the delegate only after **both** the cluster-state update and reroute callbacks complete [two callback paths](https://github.com/elastic/elasticsearch/blob/v9.4.1/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/AllocationActionListener.java#L73-L99).

4. With the desired-balance allocator, the reroute listener is queued while allocation is computed asynchronously [here](https://github.com/elastic/elasticsearch/blob/v9.4.1/server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceShardsAllocator.java#L277-L299).

5. In this occurrence, the asynchronous callback took longer than the shard restore. The restore succeeded and its `RestoreInProgress` entry was removed before the combined listener notified CCR.

6. CCR then registered `RestoreClusterStateListener`. Because the restore entry was already gone, the listener followed its explicit “too late” branch and returned a null `RestoreInfo` [here](https://github.com/elastic/elasticsearch/blob/v9.4.1/server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/restore/RestoreClusterStateListener.java#L49-L57)

7. `TransportPutFollowAction` handles a null `RestoreInfo` by returning `(true, false, false)` instead of calling `initiateFollowing` [here](https://github.com/elastic/elasticsearch/blob/v9.4.1/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java#L279-L292). The skipped method is what creates the shard-follow tasks [here](https://github.com/elastic/elasticsearch/blob/v9.4.1/x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/TransportPutFollowAction.java#L299-L322).

The index is consequently left with CCR metadata but no `ShardFollowTask` tasks, so it is reported as `paused`. A manual `resume_follow` successfully creates the missing tasks.

A possible fix would be to let CCR register its restore observer immediately after the restore cluster-state update, before waiting for asynchronous reroute completion.
